lib/rules/no-restricted-imports.js

```diff
@@ -86,7 +104,15 @@ const arrayOfStringsOrObjectPatterns = {
                     }
                 },
                 additionalProperties: false,
-                required: ["group"]
+                required: ["group"],
+                not: {
+                    anyOf: [
+                        { required: ["importNames", "allowImportNames"] },
+                        { required: ["importNamePattern", "allowImportNamePattern"] },
+                        { required: ["importNames", "allowImportNamePattern"] },
+                        { required: ["importNamePattern", "allowImportNames"] }
```

Comment:
```suggestion
                        { required: ["importNamePattern", "allowImportNames"] },
                        { required: ["allowImportNames", "allowImportNamePattern"] }
```

As currently implemented, if both `allowImportNames` and `allowImportNamePattern` are specified, then import names must satisfy both conditions. For example:

```js
/*eslint no-restricted-imports: ["error", { patterns: [{
    group: ["foo"],
    allowImportNames: ["a"]
    allowImportNamePattern: "^b",
}]}]*/

import {
    a, // 'a' import from 'foo' is restricted because only imports that match the pattern '/^b/u' are allowed.
    b  // 'b' import from 'foo' is restricted because only 'a' import(s) is/are allowed.
} from 'foo';

```

I think this doesn't make sense because specifying `allowImportNamePattern` in addition to `allowImportNames` seems redundant if both must be satisfied. We could change this either by allowing import names that satisfy at least one of the two conditions, which I think would be the expected behavior if it is allowed to specify both options but might make error messages more complicated, or by disallowing specifying both options at the same time, which I think is fine.